backend/src/App.ts

```diff
@@ -0,0 +1,96 @@
+import dotenv from "dotenv";
+import express, { Application } from "express";
+import cors from "cors";
+import bodyParser from "body-parser";
+import routes from "./routes/index";
+import { MongoClient } from "mongodb";
+import DbUpdateService from "./services/DbUpdateService";
+import CMSAuthService from "./services/CMSAuthService";
+import http from "http";
+
+/**
+ * App wrapper
+ *
+ * @class
+ * @author ale8k
+ */
+export default class App {
+    /**
+     * @constant {Application} _app Express instance
+     * Note, needs to be public so we can access the instance for
+     * supertest.
+     */
+    public readonly APP: Application;
+    /**
+     * Server references
+     */
+    public server: http.Server;
+
+    /**
+     * Registers:
+     *  - ENV variables
+     *  - App instance
+     */
+    constructor(configPath: { path: string }) {
+        console.log("Setting up environment...");
+        dotenv.config(configPath);
+        this.APP = express();
+    }
+
+    /**
+     * Server initialisation
+     */
+    public async setupServer(): Promise<Application> {
+        return await new Promise<Application>(res => {
+            console.log("Setting up middleware...");
+            this.APP.use(cors());
+            this.APP.use(bodyParser.urlencoded({
+                extended: true
+            }));
+            this.APP.use(bodyParser.json());
+            this.registerRoutes();
+
+            res(new Promise(async resolve => {
+                console.log("Setting up db connection...");
+                    await MongoClient.connect((process.env.DB_URL as string),
+                    {
+                        poolSize: 10,
+                        useNewUrlParser: true,
+                        useUnifiedTopology: true
+                    },
+                    (err, client) => {
+                        if (err) {
+                            throw err;
+                        } else {
+                            CMSAuthService.db = client.db(process.env.DB_NAME as string);
+                            console.log("CMSAuthService db reference stored...");
+                            DbUpdateService.db = client.db(process.env.DB_NAME as string);
+                            console.log("DbUpdateService db reference stored...");
+                            console.log("Resolving server...");
+                            resolve(this.APP);
+                        }
+                    }
+                );
+            }));
```

Comment:
I'm so confused with this one. Looks fine but is there any way we could've achieved it effectively with the use of `Promise.all()`

Reply:
Almost yeah, but promise.all doesn't force an order, instead it let's us perform a callback after *all* have been resolved. So we can Promise.all(). There's one issue though, when the DB connects we resolve the wrapper promise, without this we got no way to ensure the db sets in time before we return the .THIS.APP and if that resolves too late, then jest will think we're importing stuff after tests run. This was the only solution I could think of honestly. Try fiddle with it and if you come up with better one, let me know and we adjust this.
```
Promise.all([new Promise(res => {
    setTimeout(() => {
        console.log("hi")
        res();
    }, 10000)
}),
new Promise(res => {
    setTimeout(() => {
        console.log("hi from second")
        res();
    }, 5000)
})])
Promise {<pending>}
VM405:9 hi from second
VM405:3 hi
```
Notice the order doesn't matter? ^ This is the first problem I mentioned.

Reply:
Oh if order matters then promise.all wont help. Would be great to use rxjs here :P tbh.

Reply:
That's a fair point actually lol. Maybe if we get everyone comfortable, we could pipe all the dependency environments into an observable XD